text/0655-enhanced-l1.md

```diff
@@ -0,0 +1,237 @@
+    ...
+}
+```
+Anytime the bucket needs to be referenced in another construct it will take in IBucket
+```
+export interface CodePipelineProps {
+    ...
+    readonly artifactBucket?: s3.IBucket;
+}
+```
+
+L1s on the other hand don’t have any object references and just take strings to refer to other objects (names, ARNs, ids etc). They also don’t implement resource interfaces, thus cannot be passed where L2s are expected.
+
+### Proposed solution
+
+To make L1s and L2s interoperable we first need to emit a minimal interface in the L1s file with a change to the generation script. This interface should include the refs attributes for that resource (marked as `primaryIdentifiers` in the schema) and any arn attributes. 
+
+e.x for the S3Bucket L1
+```
+export interface ICfnBucket {
+    attrBucketName ...
+    attrArn ...
+}
+```
+In the L2’s resource interface we can extend this L1 interface. We can then modify all references of the L2 interface to now take in the new L1 interface. Any location that accepted `IBucket` would now accept `ICfnBucket`
+
+e.x in the Bucket L2
+```
+export interface IBucket extends ICfnBucket, IResource {
+...
+}
+```
+Any location that accepted `IBucket` would now accept `ICfnBucket`
+```
+export interface CodePipelineProps {
+    ...
+    readonly artifactBucket?: s3.ICfnBucket;
+}
+```
+If those locations use a functionality from the L2 interface, we can do an upconversion to an `IBucket` by implementing an API that converts an `ICfn<Resource>` into an `I<Resource>`. We do a similar operation when a resource that isn’t created in the current CDK stack is referenced. This needs to done to maintain backwards compatibility and not break existing customers.
+
+e.x
+```
+// takes in ICfnBucket and returns IBucket
+public static fromCFNBucket(bucket: ICfnBucket): IBucket {
+...
+}
+
+// use case
+...
+private readonly bucket: s3.IBucket;
+constructor(bucket: s3.ICfnBucket) {
+    super();
+    this.bucket = s3.Bucket.fromCFnBucket(bucket);
+}
+...
+```
+Note that we won’t be able to change `I<Resource>` to `ICfn<Resource>` everywhere. Public properties with the type `I<Resource>` can’t be changed as developers might be using an `I<Resource>` specific property. We can work on deprecating these properties and creating new `_v2` version of the property that will accept the CFN interface
+
+### POC
+
+There were some internal experiments done around this idea before. We concluded that generating `ICfn<Resource>` is fairly straight forward but actually backporting the interface to our existing L2 API is manual and runs the risk of accidentally creating breaking changes e.x turning a public property `IBucket` into `ICfnBucket` (since users could be referencing `IBucket` -specific properties) or renaming a parameter to an API can cause breaking changes in Python. If we move forward with this we'll continue the work from the experiments while keeping the findings in mind.
```

Comment:
Cautiously optimistic about this approach, and definitely curious about the findings.

Reply:
You can look at this branch for a deeper look on this POC: https://github.com/aws/aws-cdk/tree/conroy/generate